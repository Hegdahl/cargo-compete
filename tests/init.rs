use cargo_compete::{shell::Shell, Opt};
use duct::cmd;
use ignore::WalkBuilder;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;
use std::{path::Path, str};
use structopt::StructOpt as _;

#[test]
fn no_crate() -> anyhow::Result<()> {
    let (output, tree) = run("\n1\n")?;
    assert_snapshot!("no_crate_output", output);
    assert_json_snapshot!("no_crate_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[test]
fn use_crate() -> anyhow::Result<()> {
    let (output, tree) = run("\n2\n")?;
    assert_snapshot!("use_crate_output", output);
    assert_json_snapshot!("use_crate_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[test]
fn use_crate_via_bianry() -> anyhow::Result<()> {
    let (output, tree) = run("\n3\n")?;
    assert_snapshot!("use_crate_via_bianry_output", output);
    assert_json_snapshot!("use_crate_via_bianry_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(input: &'static str) -> anyhow::Result<(String, serde_json::Value)> {
    let workspace = tempfile::Builder::new()
        .prefix("cargo-compete-test-workspace")
        .tempdir()?;

    let (output_file, output) = tempfile::Builder::new()
        .prefix("cargo-compete-test-output")
        .tempfile()?
        .into_parts();

    println!("{}", cmd!("git", "init", workspace.path()).read()?);

    let Opt::Compete(opt) = Opt::from_iter_safe(&["", "compete", "i"])?;

    cargo_compete::run(
        opt,
        cargo_compete::Context {
            cwd: workspace.path().to_owned(),
            shell: &mut Shell::from_read_write(Box::new(input.as_bytes()), Box::new(output_file)),
        },
    )?;

    let output_masked = std::fs::read_to_string(&output)?
        .replace(workspace.path().to_str().unwrap(), "{{ cwd }}")
        .replace(std::path::MAIN_SEPARATOR, "{{ separator }}");

    let tree = tree(workspace.as_ref())?;

    workspace.close()?;
    output.close()?;

    Ok((output_masked, tree))
}

fn tree(path: &Path) -> anyhow::Result<serde_json::Value> {
    let mut tree = serde_json::Map::new();

    for entry in WalkBuilder::new(path)
        .git_ignore(false)
        .sort_by_file_name(Ord::cmp)
        .build()
    {
        let entry = entry?;

        let components = entry
            .path()
            .strip_prefix(path)?
            .iter()
            .map(|p| p.to_str().unwrap())
            .collect::<Vec<_>>();

        let mut tree = &mut tree;
        if entry.path().is_dir() {
            for component in components {
                tree = tree
                    .entry(component)
                    .or_insert_with(|| json!({}))
                    .as_object_mut()
                    .unwrap();
            }
        } else if let [components @ .., file_name] = &*components {
            for &component in components {
                tree = tree
                    .entry(component)
                    .or_insert_with(|| json!({}))
                    .as_object_mut()
                    .unwrap();
            }
            tree.insert(
                (*file_name).to_owned(),
                json!(std::fs::read_to_string(entry.path())?),
            );
        } else {
            panic!();
        }
    }

    Ok(serde_json::Value::Object(tree))
}
