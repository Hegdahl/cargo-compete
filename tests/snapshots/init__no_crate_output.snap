---
source: tests/init.rs
expression: output
---
Websites you compete in:
1 AtCoder
2 Codeforces
3 yukicoder
Space-delimited numbers (defaults to all): Do you use crates on AtCoder?
1 No
2 Yes
3 Yes, but I submit base64-encoded programs
Number:        Wrote {{ cwd }}{{ separator }}.gitignore
       Wrote {{ cwd }}{{ separator }}atcoder{{ separator }}Cargo.toml
       Wrote {{ cwd }}{{ separator }}atcoder{{ separator }}rust-toolchain
       Wrote {{ cwd }}{{ separator }}atcoder{{ separator }}compete.toml
     Created `cargo-compete-template` package at {{ cwd }}{{ separator }}atcoder{{ separator }}cargo-compete-template
    Updating {{ cwd }}{{ separator }}atcoder{{ separator }}Cargo.lock
       Wrote {{ cwd }}{{ separator }}codeforces{{ separator }}Cargo.toml
       Wrote {{ cwd }}{{ separator }}codeforces{{ separator }}rust-toolchain
       Wrote {{ cwd }}{{ separator }}codeforces{{ separator }}compete.toml
     Created `cargo-compete-template` package at {{ cwd }}{{ separator }}codeforces{{ separator }}cargo-compete-template
    Updating {{ cwd }}{{ separator }}codeforces{{ separator }}Cargo.lock
       Wrote {{ cwd }}{{ separator }}yukicoder{{ separator }}Cargo.toml
       Wrote {{ cwd }}{{ separator }}yukicoder{{ separator }}rust-toolchain
       Wrote {{ cwd }}{{ separator }}yukicoder{{ separator }}compete.toml
     Created `cargo-compete-template` package at {{ cwd }}{{ separator }}yukicoder{{ separator }}cargo-compete-template
    Updating {{ cwd }}{{ separator }}yukicoder{{ separator }}Cargo.lock

